src/particles/diagnostics/NonlinearLensInvariants.H

```diff
@@ -0,0 +1,131 @@
+/* Copyright 2022 Chad Mitchell, Axel Huebl
+ *
+ * This file is part of ImpactX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef IMPACTX_INVARIANTS_H
+#define IMPACTX_INVARIANTS_H
+
+#include "particles/ImpactXParticleContainer.H"
+
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+
+#include <cmath>
+
+namespace impactx::diagnostics
+{
+    /** Compute invariants within the nonlinear magnetic insert element
+     *
+     *  Invariants associated with a single short segment of the
+     *  nonlinear magnetic insert described by V. Danilov and
+     *  S. Nagaitsev, PRSTAB 13, 084002 (2010), Sect. V.A.
+     */
+    struct NonlinearLensInvariants
+    {
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        /** Independent coordinates
+         *
+         * Independent coordinates that are tracked as invariants of motion.
+         */
+        struct Data {
+            amrex::ParticleReal H; ///< first generalized coordinate
+            amrex::ParticleReal I; ///< second generalized coordinate
+        };
```

Comment:
Please double check if the documentation is descriptive enough.

Reply:
I will modify this documentation and push a new commit soon.